source/common/router/route_config_update_receiver_impl.cc

```diff
@@ -19,35 +19,63 @@ bool RouteConfigUpdateReceiverImpl::onRdsUpdate(
   if (new_hash == last_config_hash_) {
     return false;
   }
-  route_config_proto_ = rc;
+  route_config_proto_ = std::make_unique<envoy::config::route::v3::RouteConfiguration>(rc);
   last_config_hash_ = new_hash;
   const uint64_t new_vhds_config_hash = rc.has_vhds() ? MessageUtil::hash(rc.vhds()) : 0ul;
   vhds_configuration_changed_ = new_vhds_config_hash != last_vhds_config_hash_;
   last_vhds_config_hash_ = new_vhds_config_hash;
-  initializeRdsVhosts(route_config_proto_);
-  onUpdateCommon(route_config_proto_, version_info);
-  return true;
-}
+  initializeRdsVhosts(*route_config_proto_);
 
-void RouteConfigUpdateReceiverImpl::onUpdateCommon(
-    const envoy::config::route::v3::RouteConfiguration& rc, const std::string& version_info) {
-  last_config_version_ = version_info;
-  last_updated_ = time_source_.systemTime();
-  rebuildRouteConfig(rds_virtual_hosts_, vhds_virtual_hosts_, route_config_proto_);
-  config_info_.emplace(RouteConfigProvider::ConfigInfo{rc, last_config_version_});
+  rebuildRouteConfig(rds_virtual_hosts_, *vhds_virtual_hosts_, *route_config_proto_);
+  config_ = std::make_shared<ConfigImpl>(
+      *route_config_proto_, factory_context_,
+      factory_context_.messageValidationContext().dynamicValidationVisitor(), false);
+
+  onUpdateCommon(version_info);
+  return true;
 }
 
 bool RouteConfigUpdateReceiverImpl::onVhdsUpdate(
     const VirtualHostRefVector& added_vhosts, const std::set<std::string>& added_resource_ids,
     const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& version_info) {
+
+  auto vhosts_before_this_update =
+      std::make_unique<std::map<std::string, envoy::config::route::v3::VirtualHost>>(
+          *vhds_virtual_hosts_);
+  const bool removed = removeVhosts(*vhds_virtual_hosts_, removed_resources);
+  const bool updated = updateVhosts(*vhds_virtual_hosts_, added_vhosts);
+
+  auto route_config_before_this_update =
+      std::make_unique<envoy::config::route::v3::RouteConfiguration>();
+  route_config_before_this_update->CopyFrom(*route_config_proto_);
+  rebuildRouteConfig(rds_virtual_hosts_, *vhds_virtual_hosts_, *route_config_proto_);
+  ConfigConstSharedPtr new_config;
+
+  try {
+    new_config = std::make_shared<ConfigImpl>(
+        *route_config_proto_, factory_context_,
+        factory_context_.messageValidationContext().dynamicValidationVisitor(), false);
+  } catch (const Envoy::EnvoyException& e) {
+    // revert the changes that failed validation
+    vhds_virtual_hosts_ = std::move(vhosts_before_this_update);
+    route_config_proto_ = std::move(route_config_before_this_update);
```

Comment:
Attempting to revert on exception seems a bit brittle.  Would it be possible to avoid doing the update in the first place until we are sure validation succeeded?  Construct new inputs, and move into vhds_virtual_hosts_ / route_config_proto_ on success.

Reply:
> Would it be possible to avoid doing the update in the first place until we are sure validation succeeded?

I think so, I'll do another PR.

Reply:
That's a good idea, thanks @antoniovicente.
@dmitri-d let me know if you'd like me to work on this.

Reply:
Filed #15320 to track this.

Reply:
> @dmitri-d let me know if you'd like me to work on this.

Cheers, @adisuissa, I should be able to get to this tomorrow, will ping you here should things get away from me.